Adding DMStag functions as well as testing and documentation

nicoberlie opened the pull request:
- Adding DMStag functionality into `PETSc.jl`
- Adding documentation to newly introduced functions
- Adding testing for additional functions
- Adding examples using DMStag and Automatic Differentiation tools

jkozdon commented:
Thanks! Will take a look soon (tomorrow?)

I do notice that there are some conflicts with some of the changes I've recently put onto master.

Can you pull the latest master and merge?

jkozdon commented:
(Let me know if you need help with this)

nicoberlie commented:
Hi, I just did that, I think it solved most conflicts.

jkozdon commented:
> Hi, I just did that, I think it solved most conflicts.

Thanks! I'll take a look later today.

jkozdon commented on src/snes.jl:
```diff
@@ -12,8 +12,11 @@ mutable struct SNES{T}
     update_jac!
     jac_A
     jac_P
+    julia_vec::Cint 
```
Maybe use a boolean?
```suggestion
    use_julia_vec::Bool 
```

jkozdon commented on test/Project.toml:
```diff
@@ -0,0 +1,8 @@
+[deps]
+ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
+Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+MPI = "da04e1cc-30fd-572f-bb4f-1f8673147195"
+PETSc_jll = "8fa3689e-f0b9-5420-9873-adf6ccf46f2d"
+SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
```
Does the test directory need a `Project.toml`?

boriskaus replied:
ForwardDiff and SparseArrays (and Plots) are dependencies that we only need for certain tests or examples, but is not strictly speaking necessary for PETSc.jl itself. I understood that this is the way to do that. Fewer dependencies on the main package will likely make it easier to maintain.

jkozdon replied:
I think that you can put `test` dependencies in the `extra` section of the main `Project.toml`. But maybe putting them here is an alternative approach?

boriskaus replied:
I did put all under `extra` now & deleted the `/test/Project.toml` file.

jkozdon reviewed (commented):
I've made a quick first pass through. I didn't go through all the examples.

Couple questions I have are in the comments. Main change requested is to fix the doc strings (BTW, I love that there are docs!)
 
We'll need to setup the documentation deployment too, but I can fix that.

Also, we need to either at `Plots` to the project toml `extras` section (or switch to `unicode` plots for the plotting). The nice thing about `unicode` plots is that we don't need our testing to depending on the heavy plotting package, but I'm ultimately fine either way.)

jkozdon commented:
BTW, this is awesome! Thanks for adding the functionality

jkozdon commented on src/dmstag.jl:
```diff
@@ -0,0 +1,1308 @@
+
+@for_libpetsc begin
+
+    """
+        dm = DMStagCreate1d(comm::MPI.Comm, bndx::DMBoundaryType, M, dofVertex, dofCenter, stencilType::DMStagStencilType=DMSTAG_STENCIL_BOX, stencilWidth=2, lx=C_NULL; kwargs...)
+
+    Creates a 1D DMStag object.
+
+            comm            -   MPI communicator
+            bndx            -   boundary type: DM_BOUNDARY_NONE, DM_BOUNDARY_PERIODIC, or DM_BOUNDARY_GHOSTED. 
+            M               -   global number of grid points
+            dofVertex       -   [=1] number of degrees of freedom per vertex/point/node/0-cell
+            dofCenter       -   [=1] number of degrees of freedom per element/edge/1-cell
+            stencilType     -   ghost/halo region type: DMSTAG_STENCIL_BOX or DMSTAG_STENCIL_NONE
+            stencilWidth    -   width, in elements, of halo/ghost region
+            lx              -   [Optional] Vector of local sizes, of length equal to the comm size, summing to M
+            kwargs...       -   [Optional] keyword arguments (see PETSc webpage), specifiable as stag_grid_x=100, etc. 
+    """
+    function DMStagCreate1d(comm::MPI.Comm, bndx::DMBoundaryType, M, dofVertex=1,dofCenter=1,stencilType::DMStagStencilType=DMSTAG_STENCIL_BOX,stencilWidth=2, lx=C_NULL; kwargs...)
```
Another docs comment, I wondering if we should add links back to the PETSc docs whenever possible. (I would say the new petsc website: https://petsc.org/release/documentation/manualpages/)

boriskaus replied:
Regarding the docs: I started another [branch](https://github.com/boriskaus/PETSc.jl/tree/documentation_setup) that has documentation. I will merge the DMSTAG part into that & set up the pipeline. It's indeed a good point to add links to the PETSc manual. I did go over this (julia make.jl in docs results in OK formatted function descriptions).

I guess we didn't really understand what the @for_petsc is doing; I'm happy to add this to very function. 

As to Plots: I'm also fine if we replace that. It is indeed a heavy package. We only use it in the examples.

jkozdon replied:
So the `@for_petsc` macro make sure that we link against the right petsc shared library. So have the docs inside the macro, means that we will generate the same docs for each version of the function. So you will end up with something like:

```
julia> macro for_IT(expr)
         quote
           for IT in (Int32, Int64)
             @eval esc($expr)
           end
         end
       end
@for_IT (macro with 1 method)

julia> @for_IT begin
         """
             foo(a)

         some docs
         """
         function foo(a::$IT)
           return a + 1
         end
       end

help?> foo
search: foo floor pointer_from_objref OverflowError RoundFromZero unsafe_copyto! functionloc StackOverflowError @functionloc OutOfMemoryError UndefKeywordError

  foo(a)

  some docs

  ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

  foo(a)

  some docs
```

And the function `foo` shows up twice in the docs.

jkozdon replied:
I'm fine with fixing the docs in another PR if it would be better to just get this merged first.

boriskaus commented on src/dmstag.jl:
```diff
@@ -0,0 +1,1308 @@
+
+        v = VecSeq(C_NULL, dm.comm, [0.0])  # empty vector
+        
+        ccall((:DMGetCoordinatesLocal, $libpetsc), PetscErrorCode, (CDMStag, Ptr{CVec}), dm, v)
```
all fine with me

boriskaus commented on src/dmstag.jl:
```diff
@@ -0,0 +1,1308 @@
+const CDMStag = Ptr{Cvoid}
+const CDMStagType = Cstring
+
+mutable struct DMStag{T} <: Factorization{T}
+    ptr::CDMStag
+    comm::MPI.Comm
+    dim::Int64
+    opts::Options{T}
+end
+
+mutable struct DMSTAGSTENCIL
+    loc::DMStagStencilLocation
+    i::Int64
+    j::Int64
+    k::Int64
+    c::Int64
+end
+
+mutable struct DMSTAGSTENCIL_C
+    loc::DMStagStencilLocation
+    i::Cint
+    j::Cint
+    k::Cint
+    c::Cint
+end
+
+const DMStagStencil     = DMSTAGSTENCIL
+const DMStagStencil_c   = DMSTAGSTENCIL_C
```
ok, we did not have this on our radar when writing, also as only one version of PETSc_jll was linked.

nicoberlie replied:
Hi @jkozdon this part with the structures is still a bit unclear to us. I managed to use `PetscInt` into the `DMStag{T,PetscInt}` structure (it's been added to the PR already, you can have a look) but it doesn't seem to be as simple for the `DMStagStencil `and `DMStagStencil_c` ones. Could you give me a hand for that?  I guess it's a bit different as the structure `DMStagStencil_c` is meant as an input for a Petsc function?

jkozdon replied:
I'll take a look right now.

nicoberlie commented on src/dmstag.jl:
```diff
@@ -0,0 +1,1308 @@
+        #Arry = C_NULL
+        #Arrz = C_NULL
+
+        ccall((:DMStagGetProductCoordinateArraysRead, $libpetsc), PetscErrorCode, (CDMStag, Ptr{$PetscScalar}, Ptr{$PetscScalar}, Ptr{$PetscScalar}), dm, Arrx, Arry, Arrz)
```
Thanks for pointing those out!

jkozdon commented on src/options.jl:
```diff
@@ -67,7 +67,9 @@ Create a new PETSc options database.
     opts = Options{T}()
     for (k,v) in ps
-        opts[k] = v
+        if  v!=false
+            opts[k] = v
+        end
```
I understand now what the goal is with this convo: https://github.com/JuliaParallel/PETSc.jl/pull/126#issuecomment-876529653

jkozdon commented:
So I think that I see some of the problems, part of the hard part right now is that we have the code setup for multiple PETSc library testing but not the CI (since we are waiting on https://github.com/JuliaPackaging/Yggdrasil/pull/3249 for updated binaries).

Locally I have setup to run various PETSc configurations (different `PetscScalar` and `PetscInt` values), and it pops a bunch of errors.

Mainly anytime you use the `@for_libpetsc` you need to have `$PetscScalar` in the function definition, so for instance
```julia
@for_libpetsc function DMStagCreate1d(::Type{$PetscScalar}, comm::MPI.Comm, bndx::DMBoundaryType, M, dofVertex=1,dofCenter=1,stencilType::DMStagStencilType=DMSTAG_STENCIL_BOX,stencilWidth=2, lx=C_NULL; kwargs...)
```
and then when you use it you would call something like
```julia
DMStagCreate1d(Float64, ....)
```
so that it knows which PETSc library to use.

For function that take an argument with type `DMStag` things are simpler since you just need to change to doing
```julia
@for_libpetsc function foo(dm::DMStag{$PetscScalar}, ...
```
(which you've done in most places already.)

I know that this is a bit of a pain, but I think that it will pay off in the future in order to add more flexibility.

Alternatively (and perhaps a better approach) is to switch to typing the `DMStag` struct on the library type and using my new `@for_petsc` macro; see for instance how I modified the [`viewer`](https://github.com/JuliaParallel/PETSc.jl/blob/03f87adfe07b9ac4bd4bb35febab343f79701b55/src/viewer.jl)

I am happy to set up a little VTC pair programming session to talk / work through things. Alternatively, I could clean up the previous DMDA stuff I started in #120 with these ideas and you could mirror what I do there.

boriskaus commented:
Thanks, that is indeed part of what we were not understanding. It would be really helpful if you update your DMDA stuff in the way you think it works best.
In addition it would be great if you can give us an example of how to do this with `DMStagVecSetValuesStencil`:
Currently we define two structs and a convert routine for that:
```julia
mutable struct DMSTAGSTENCIL
    loc::DMStagStencilLocation
    i::Int64
    j::Int64
    k::Int64
    c::Int64
end
mutable struct DMSTAGSTENCIL_C
    loc::DMStagStencilLocation
    i::Cint
    j::Cint
    k::Cint
    c::Cint
end
Base.convert(::Type{DMStagStencil_c}, v::DMStagStencil) = DMStagStencil_c(v.loc, v.i, v.j,v.k, v.c)
```
If we use different `PetscInt` values, we will obviously have to change that from `Int64` to `PetscInt`. Yet, do we have to do the same with `Cint`? 

A related question that I am not fully understanding: a different PETSc library has a specific `$PetscScalar` and a specific `$PetscInt` (32 or 64 bit). Why is it in that case sufficient to only define the `$PetscScalar`?
```julia 
DMStagCreate1d(Float64, ....)
```

jkozdon commented:
> Thanks, that is indeed part of what we were not understanding. It would be really helpful if you update your DMDA stuff in the way you think it works best.

Great. I'll work on in the afternoon.

> In addition it would be great if you can give us an example of how to do this with `DMStagVecSetValuesStencil`

I'll give this some thought as well.

> A related question that I am not fully understanding: a different PETSc library has a specific `$PetscScalar` and a specific `$PetscInt` (32 or 64 bit). Why is it in that case sufficient to only define the `$PetscScalar`?

It's not! That's just the old way the library was doing, that's why I want to move to using the [`PetscLibType`](https://github.com/JuliaParallel/PETSc.jl/blob/75333979759a1251fa06cd4e8b21ee836eab1b5c/src/lib.jl#L2-L16) since it (can) encapsulate any PETSc build options more easily. So each PETSc build gets a unique one of these types, and all the dispatch is based on it.

Right now it just uses the `PetscScalar` and `PetscInt` types, but we can easily add other build parameters (debug mode? Blas Types? etc.) and not other code needs to change.

boriskaus commented:
> Right now it just uses the `PetscScalar` and `PetscInt` types, but we can easily add other build parameters (debug mode? Blas Types? etc.) and not other code needs to change.

That would indeed be great to have! I usually work with both a debug and optimised PETSc version, with various external packages, so would be cool to have the possibility to change that in a sraightforward manner

jkozdon commented:
So I got most of the work done on my side, and I am starting to fold some of the routines that you all call from `DM` into my PR.

I still need to fold in:

- `DMCreateGlobalVector`
- `DMCreateLocalVector`
- `DMLocalToGlobal`
- `DMGlobalToLocal`
- `DMCreateMatrix`
- `DMGetCoordinateDM`
- `DMGetCoordinatesLocal`

Concerning `DMStagVecSetValuesStencil` I realized that I did a similar thing with [`MatStencil`](https://github.com/JuliaParallel/PETSc.jl/blob/3145ea1556bd893eb08774e4f57c155384dae78e/src/mat.jl#L52-L94) and [`MatSetValuesStencil `](https://github.com/JuliaParallel/PETSc.jl/blob/3145ea1556bd893eb08774e4f57c155384dae78e/src/mat.jl#L137-L153) (and [here](https://github.com/JuliaParallel/PETSc.jl/blob/3145ea1556bd893eb08774e4f57c155384dae78e/src/mat.jl#L227-L249))

boriskaus commented:
Thanks! Yes, putting the `DM` routines into a separate file makes sense.

nicoberlie commented:
All the tests now work. I reverted back on SNES to give back the vectors' pointers as I thought it would be more explicit to transform it to an array in the exposed code (especially if it's associated to a dm). I also added a user context to the snes structure as it can be useful to carry some information in it. We can discuss those points if you want. All the other changes were fixes.

I know there is more changes that could be added but, at this point and knowing @jkozdon is working on an automatically generated wrapper, could we add this to the code and eventually add subsequent pull requests?

Enjoy your holiday, in the mean time I'll have a look at the structure `DMSTAGSTENCIL`

boriskaus commented:
All tests work fine now so in principle this could be merged now.
Yet, compared to different parts of the package, we are only doing the test for one PETSc library. Perhaps it is worth adding a loop over libraries here so we test all combinations? I am not sure whether `DMSTAG` is even supposed to work with complex numbers, though.

nicoberlie commented:
That's true I also had troubles with other petsc libraries so I came back to the one we had before in the startup file, maybe we can discuss it Jeremy?

________________________________
De : Boris Kaus ***@***.***>
Envoyé : mardi 27 juillet 2021 16:58:29
À : JuliaParallel/PETSc.jl
Cc : Berlie, Nicolas; Author
Objet : Re: [JuliaParallel/PETSc.jl] Adding DMStag functions as well as testing and documentation (#119)


All tests work fine now so in principle this could be merged now.
Yet, compared to different parts of the package, we are only doing the test for one PETSc library. Perhaps it is worth adding a loop over libraries here so we test all combinations? I am not sure whether DMSTAG is even supposed to work with complex numbers, though.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/JuliaParallel/PETSc.jl/pull/119#issuecomment-887585162>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AT5VXZ7WKYPYJNDHWOYSVG3TZ3CRLANCNFSM472I5SAA>.

jkozdon commented:
> That's true I also had troubles with other petsc libraries so I came back to the one we had before in the startup file, maybe we can discuss it Jeremy?

Sure we can discuss next week. I'll be back in the office Monday and can chat then.

nicoberlie commented:
Thanks, I'll have a couple questions on that matter.


Nico

________________________________
De : Jeremy E Kozdon ***@***.***>
Envoyé : mercredi 28 juillet 2021 04:51:28
À : JuliaParallel/PETSc.jl
Cc : Berlie, Nicolas; Author
Objet : Re: [JuliaParallel/PETSc.jl] Adding DMStag functions as well as testing and documentation (#119)


That's true I also had troubles with other petsc libraries so I came back to the one we had before in the startup file, maybe we can discuss it Jeremy?

Sure we can discuss next week. I'll be back in the office Monday and can chat then.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/JuliaParallel/PETSc.jl/pull/119#issuecomment-887968545>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AT5VXZ3D6WUBDQZZ7EG6GRTTZ5WDBANCNFSM472I5SAA>.
